### ISSUE TITLE
[question] transitivity of package_id() declarations 

### ISSUE THREAD

**Issue by michaelmaguire** [conan-io/conan#4686]
# Summary

Reading through https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html#versioning-schema, can I double check that versioning schema declarations respect transitive dependencies?

# Example

 * `A` depends on `B`
 * `A` depends on `C`
 * `B` depends on `D`
 * `C` depends on `D`
 * `A` *does not directly depend* on `D`

we set in package `B` and `C` something like:

```
	def package_id(self):
		self.info.requires["D"].minor_mode()
```

we have pre-built packages:

`D`  v5.8 is built
`C`  requires `D` [>=5.0], and `C` v1.7 is built using `D` v5.8
`B`  requires `D` [>=5.0], and `B` v3.4 is built using `D` v5.8

`A` requires `B` [>=3.0], `C` [>=1.0], v1.2 built using `B` v3.4, `C` v1.7

![image](https://user-images.githubusercontent.com/2184330/53900216-ecb05680-4033-11e9-84b8-a3ad19071e19.png)

Then a new version of package `D` is built, v5.9

No new versions of `B` or `C` are built yet.

Now I rebuild `A` v1.2.  Conan looks for latest versions of `B`, `C`, finds nothing new, re-uses binaries `B` v3.4, `C` v1.7.  As expected.

Next I rebuild `B` v3.4 with *no* source changes and *no* bump to its version number, but conan picks up the latest `D` v5.9.

![image](https://user-images.githubusercontent.com/2184330/53900329-38fb9680-4034-11e9-8329-9c3c7d7ef663.png)

Does a rebuild of `A` grab the latest `B` v3.4, but examining its dependencies sees that it uses a different `D` version, so examining the current `C` v1.7 that was built against old `D` v5.8 realizes by virtue of `C`'s ```minor_mode()``` declaration that it can't use those current `C` and `B` binary packages together?

Note:  `A` itself has no package_id() declaration mentioning `D`.

```
	def package_id(self):
		self.info.requires["D"].minor_mode()
```

To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

Conan Windows  version 1.10.1

**Comment by jgsogo** [comment]
Hi @michaelmaguire,

When I follow the steps you are providing (I've named the libraries `libA`, `libB`,...) I found that, after creating `libD/5.9`, when I try to build again `libA` Conan resolves a new version of `libD` for the _version_range_ in recipes `libB` and `libC` and try to compile both of them:

```bash
Version range '>=3.0' required by 'libA/1.2@issue/testing' resolved to 'libB/3.4@issue/testing'
Version range '>=1.0' required by 'libA/1.2@issue/testing' resolved to 'libC/1.7@issue/testing'
Version range '>=5.0' required by 'libB/3.4@issue/testing' resolved to 'libD/5.9@issue/testing'
Version range '>=5.0' required by 'libC/1.7@issue/testing' resolved to 'libD/5.9@issue/testing'
libA/1.2@issue/testing: WARN: Forced build from source
libA/1.2@issue/testing: Installing package
Requirements
    libA/1.2@issue/testing from local cache - Cache
    libB/3.4@issue/testing from local cache - Cache
    libC/1.7@issue/testing from local cache - Cache
    libD/5.9@issue/testing from local cache - Cache
Packages
    libA/1.2@issue/testing:edbcf0fdd61d3be6edee6845e4da00bb8aba2cac - Build
    libB/3.4@issue/testing:1a754850f86bc95a862d33c26fe6a15d9bb9fe15 - Missing
    libC/1.7@issue/testing:1a754850f86bc95a862d33c26fe6a15d9bb9fe15 - Missing
    libD/5.9@issue/testing:f8bda7f0751e4bc3beaa6c3b2eb02d455291c8a2 - Cache

libD/5.9@issue/testing: Already installed!
libB/3.4@issue/testing: WARN: Can't find a 'libB/3.4@issue/testing' package for the specified settings, options and dependencies:
- Settings: arch=x86_64, build_type=Release, compiler=apple-clang, compiler.libcxx=libc++, compiler.version=10.0, os=Macos
- Options: shared=False, libD:shared=False
- Dependencies: libD/5.9@issue/testing
- Package ID: 1a754850f86bc95a862d33c26fe6a15d9bb9fe15

ERROR: Missing prebuilt package for 'libB/3.4@issue/testing'
Try to build it from sources with "--build libB"
Or read "http://docs.conan.io/en/latest/faq/troubleshooting.html#error-missing-prebuilt-package"
```

So, as soon as `libD/5.9` is available in the cache, every recipe depending on it with a version range will resolve to the latest version, will compute a different package ID and Conan will ask for a new binary.

I'm not sure if this was just a question, or you expected a different behavior. Thanks!

**Comment by michaelmaguire** [comment]
Thank you @jgsogo my question was hoping to confirm this was the officially supported behaviour.
